src/secdb.c

```diff
@@ -0,0 +1,847 @@
+{
+	const uint8_t *a = (const uint8_t *)ap;
+	const uint8_t *b = (const uint8_t *)bp;
+
+	for (size_t i = 0; i < sz; i++) {
+		int sub = a[i] - b[i];
+		if (sub != 0) {
+			debug("byte %zu differs: a=0x%02hhx %c b=0x%02hhx",
+			      i, a[i], LEGR(sub), b[i]);
+			return sub;
+		}
+	}
+	return 0;
+}
+
+/*
+ * compare secdb_entry_t items
+ */
+int
+secdb_entry_cmp(const void *ap, const void *bp, void *state)
+{
+	const secdb_entry_t *a = *(const secdb_entry_t **)ap;
+	const secdb_entry_t *b = *(const secdb_entry_t **)bp;
+	size_t sigsz = *(size_t *)state;
+	int rc;
+
+	rc = efi_guid_cmp(&a->owner, &b->owner);
+	if (rc != 0) {
+		debug("owner guids differ: "GUID_FORMAT" %c "GUID_FORMAT,
+		      GUID_FORMAT_ARGS(&a->owner),
+		      LEGR(rc),
+		      GUID_FORMAT_ARGS(&b->owner));
+		return rc;
+	}
+
+	return bytecmp(a->data.raw, b->data.raw, sigsz);
+}
+
+int
+secdb_entry_cmp_descending(const void *ap, const void *bp, void *state)
+{
+	return secdb_entry_cmp(bp, ap, state);
+}
+
+/*
+ * compare efi_secdb_t items
+ */
+int
+secdb_cmp(const void *ap, const void *bp, void * state UNUSED)
+{
+	const efi_secdb_t *a;
+	const efi_secdb_t *b;
+
+	if (ap == NULL || bp == NULL)
+		return ap - bp;
+
+	a = *(efi_secdb_t **)ap;
+	b = *(efi_secdb_t **)bp;
+
+	if (a->algorithm == MAX_SECDB_TYPE) {
+		debug("sorting unready data from secdb:%p", a);
+		return -1;
+	}
+
+	if (b->algorithm == MAX_SECDB_TYPE) {
+		debug("sorting unready data from secdb:%p", b);
+		return 1;
+	}
+
+	if (a->algorithm != b->algorithm) {
+		return a->algorithm - b->algorithm;
+	}
+
+	if (a->sigsz != b->sigsz) {
+		return a->sigsz - b->sigsz;
+	}
+
+	return a->listsz - b->listsz;
+}
+
+int
+secdb_cmp_descending(const void *ap, const void *bp, void * state)
+{
+	return secdb_cmp(bp, ap, state);
+}
+
+const secdb_alg_t PUBLIC efi_secdb_algs_[MAX_SECDB_TYPE] = {
+	[SHA1] = {
+		.class = HASH,
+		.guid = &efi_guid_sha1,
+		.header_size = 0,
+		.has_owner = true,
+		.size = 20
+	},
+	[SHA224] = {
+		.class = HASH,
+		.guid = &efi_guid_sha224,
+		.header_size = 0,
+		.has_owner = true,
+		.size = 28,
+	},
+	[SHA256] = {
+		.class = HASH,
+		.guid = &efi_guid_sha256,
+		.header_size = 0,
+		.has_owner = true,
+		.size = 32,
+	},
+	[SHA384] = {
+		.class = HASH,
+		.guid = &efi_guid_sha384,
+		.header_size = 0,
+		.has_owner = true,
+		.size = 48,
+	},
+	[SHA512] = {
+		.class = HASH,
+		.guid = &efi_guid_sha512,
+		.header_size = 0,
+		.has_owner = true,
+		.size = 64,
+	},
+	[RSA2048] = {
+		.class = SIGNATURE,
+		.guid = &efi_guid_rsa2048,
+		.header_size = 0,
+		.has_owner = true,
+		.size = 256,
+	},
+	[RSA2048_SHA1] = {
+		.class = SIGNATURE,
+		.guid = &efi_guid_rsa2048_sha1,
+		.header_size = 0,
+		.has_owner = true,
+		.size = 256,
+	},
+	[RSA2048_SHA256] = {
+		.class = SIGNATURE,
+		.guid = &efi_guid_rsa2048_sha256,
+		.header_size = 0,
+		.has_owner = true,
+		.size = 256,
+	},
+        [X509_SHA256] = {
+                .class = CERTIFICATE_HASH,
```

Comment:
Starting about here the whitespace gets funky - spaces/tabs thing.  It's a problem in all of the rest too except the last one.